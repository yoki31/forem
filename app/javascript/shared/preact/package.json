{
  "name": "preact-root-fragment",
  "version": "0.2.0",
  "main": "./preact-root-fragment.js",
  "module": "./preact-root-fragment.js",
  "type": "module"
}
