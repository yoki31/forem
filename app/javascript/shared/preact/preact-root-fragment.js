/**
 * A Preact 11+ implementation of the `replaceNode` parameter from Preact 10.
 *
 * This creates a "Persistent Fragment" (a fake DOM element) containing one or more
 * DOM nodes, which can then be passed as the `parent` argument to Preact's `render()` method.
 * Source: https://gist.github.com/developit/f4c67a2ede71dc2fab7f357f39cff28c
 */
export function createRootFragment(parent, replaceNode) {
  if (replaceNode) {
    replaceNode = Array.isArray(replaceNode) ? replaceNode : [replaceNode];
  } else {
    replaceNode = [parent];
    parent = parent.parentNode;
  }

  const s = replaceNode[replaceNode.length - 1].nextSibling;

  const rootFragment = {
    nodeType: 1,
    parentNode: parent,
    firstChild: replaceNode[0],
    childNodes: replaceNode,
    insertBefore: (c, r) => {
      parent.insertBefore(c, r || s);
      return c;
    },
    appendChild: (c) => {
      parent.insertBefore(c, s);
      return c;
    },
    removeChild: (c) => {
      parent.removeChild(c);
      return c;
    },
  };

  parent.__k = rootFragment;
  return rootFragment;
}
