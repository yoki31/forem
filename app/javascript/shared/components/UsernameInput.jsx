import { h } from 'preact';
import PropTypes from 'prop-types';
import { MultiSelectAutocomplete } from '@crayons/MultiSelectAutocomplete/MultiSelectAutocomplete';

/**
 * UsernameInput â€” produces a field that can autocomplete usernames
 *
 * @param {Function} fetchSuggestions Callback to sync selections to article form state
 * @param {string} defaultValue Comma separated list of any currently user IDs
 */
export const UsernameInput = ({
  fetchSuggestions,
  defaultValue,
  inputId,
  labelText,
  placeholder,
  maxSelections,
  handleSelectionsChanged,
}) => {
  const onSelectionsChanged = function (selections) {
    const ids = selections.map((item) => item.id).join(', ');
    handleSelectionsChanged?.(ids);
  };

  return (
    <MultiSelectAutocomplete
      allowUserDefinedSelections={false}
      showLabel={false}
      border={true}
      inputId={inputId}
      labelText={labelText}
      placeholder={placeholder}
      maxSelections={maxSelections}
      defaultValue={defaultValue}
      fetchSuggestions={fetchSuggestions}
      onSelectionsChanged={onSelectionsChanged}
    />
  );
};

UsernameInput.propTypes = {
  fetchSuggestions: PropTypes.func.isRequired,
  defaultValue: PropTypes.string,
  inputId: PropTypes.string,
  labelText: PropTypes.string,
  placeholder: PropTypes.string,
  maxSelections: PropTypes.string,
  handleSelectionsChanged: PropTypes.func.isRequired,
};
