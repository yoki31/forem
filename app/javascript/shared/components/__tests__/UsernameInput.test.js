import { h } from 'preact';
import { render } from '@testing-library/preact';
import { userEvent } from '@testing-library/user-event';

import '@testing-library/jest-dom';

import { UsernameInput } from '../UsernameInput';

jest.mock('@utilities/debounceAction', () => ({
  debounceAction: fn => fn
}));

function fakeUsers() {
  return [
    { name: 'Alice', username: 'alice', id: 1 },
    { name: 'Bob', username: 'bob', id: 2 },
    { name: 'Charlie', username: 'charlie', id: 3 },
    { name: 'Almost Alice', username: 'almostalice', id: 4 },
  ];
}

describe('<UsernameInput />', () => {
  it('renders the default component', () => {
    const { container } = render(
      <UsernameInput
        labelText="Example label"
        fetchSuggestions={() => {}}
        handleSelectionsChanged={() => {}}
      />,
    );
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('calls handleSelectionsChanged with user IDs', async () => {
    const fakeHandler = jest.fn();

    const { getByLabelText, queryByRole } = render(
      <UsernameInput
        labelText="Enter username"
        fetchSuggestions={fakeUsers}
        handleSelectionsChanged={fakeHandler}
      />,
    );

    const input = getByLabelText('Enter username');
    input.focus();
    await userEvent.type(input, 'Bob,');
    await userEvent.type(input, 'Charlie,');

    expect(queryByRole('button', { name: 'Edit example' })).toBeNull();
    expect(fakeHandler).toHaveBeenCalledWith('2, 3');
  });
});
