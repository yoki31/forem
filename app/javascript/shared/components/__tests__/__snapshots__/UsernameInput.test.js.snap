// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UsernameInput /> renders the default component 1`] = `"<span aria-hidden=\\"true\\" class=\\"absolute pointer-events-none opacity-0 p-2\\"></span><label id=\\"multi-select-label\\" class=\\"screen-reader-only\\">Example label</label><span id=\\"input-description\\" class=\\"screen-reader-only\\"></span><div class=\\"screen-reader-only\\"><p>Selected items:</p><ul aria-live=\\"assertive\\" aria-atomic=\\"false\\" aria-relevant=\\"additions removals\\" class=\\"screen-reader-only list-none\\"></ul></div><div class=\\"c-autocomplete--multi relative\\"><div role=\\"combobox\\" aria-haspopup=\\"listbox\\" aria-expanded=\\"false\\" aria-owns=\\"listbox1\\" class=\\"c-autocomplete--multi__wrapper-border crayons-textfield flex items-center  cursor-text\\"><ul id=\\"combo-selected\\" class=\\"list-none flex flex-wrap w-100\\"><li style=\\"order: 1;\\" class=\\"self-center\\"><input autocomplete=\\"off\\" aria-autocomplete=\\"list\\" aria-labelledby=\\"multi-select-label selected-items-list\\" aria-describedby=\\"input-description\\" aria-disabled=\\"false\\" type=\\"text\\" placeholder=\\"Add...\\" class=\\"c-autocomplete--multi__input\\"></li></ul></div></div>"`;
